library/FrontRow/App/Http.hs

```diff
@@ -0,0 +1,122 @@
+-- | Centralized module for making HTTP from the backend
+--
+-- These functions:
+--
+-- - Do not throw exceptions on non-200
+-- - May throw for other 'HttpException' cases (e.g. 'ConnectionTimeout')
+-- - Handle 429-@Retry-In@ for you
+-- - Capture decoding failures with 'Either' values as the 'Response' body
+--
+module FrontRow.App.Http
+  ( httpJson
+  , HttpDecodeError(..)
+  , httpDecode
+  , httpLbs
+  , httpPaginated
+  , sourcePaginated
+
+  -- * Request builders
+  , Request
+  , parseRequest
+  , parseRequest_
+
+  -- * Request modifiers
+  , addRequestHeader
+  , addAcceptHeader
+  , addBearerAuthorizationHeader
+  , addToRequestQueryString
+  , setRequestBasicAuth
+
+  -- * Response accessors
+  , Response
+  , getResponseStatus
+  , getResponseBody
+
+  -- ** Unsafe access
+  , getResponseBodyUnsafe
+
+  -- * "Network.HTTP.Types" re-exports
+  , Status
+  , statusCode
+  )
+where
+
+import Prelude
+
+import Conduit (foldC, mapMC, runConduit, (.|))
+import Control.Monad.IO.Class (MonadIO)
+import Data.Aeson (FromJSON)
+import qualified Data.Aeson as Aeson
+import Data.Bifunctor (first)
+import qualified Data.ByteString as BS
+import Data.ByteString.Lazy (ByteString)
+import qualified Data.ByteString.Lazy.Char8 as BSL8
+import FrontRow.App.Http.Paginate
+import FrontRow.App.Http.Retry
+import Network.HTTP.Simple hiding (httpLbs)
+import Network.HTTP.Types.Header (hAccept, hAuthorization)
+import Network.HTTP.Types.Status (Status, statusCode)
+import UnliftIO.Exception (Exception(..), throwIO)
+
+data HttpDecodeError = HttpDecodeError
+  { hdeBody :: ByteString
+  , hdeErrors :: String
```

Comment:
Are there actually multiple errors embedded in one string here?

Reply:
With `Aeson.eitherDecode` that is true, but that's just one use-case.

From the perspective of the library, it's just leaving the function you pass to `httpDecode` in charge of formatting whatever problem or problems it generates into a single message, which I think is a useful separate of responsibility.

Reply:
Right, but `httpDecode` doesn't require that the `decode` function return `Either HttpDecodeError a`, only `Either e a`. Since we're only using it for the JSON case, is there a reason to not make this error JSON-specific? Or is the expectation that clients will also want to use `HttpDecodeError`? If the only point of `HttpDecodeError` is to bundle _some_ error with the body, we could do:

```haskell
data HttpDecodeError e = HttpDecodError
  { hdeBody :: ByteString
  , hdeErrors :: NonEmpty e
  }
  deriving (Eq, Show)

instance Exception e => Exception (HttpDecodeError e) where
  displayException HttpDecodeError {..} = intercalate
    "\n"
    [ "Failed to decode HTTP Response"
    , "Raw body: {"
    , BSL8.unpack hdeBody
    , "}"
    , fromErrors hdeErrors
    ]
   where
    fromErrors = \case
      err :| [] -> "Error:\n" <> displayException e
      errs ->
        "Errors:\n" <> intercalate "\n" (bullet . displayException <$> errs)
    bullet = (" • " <>)
```

Reply:
I like this, but it's starting to untangle on me. I ended up at:

```hs
-- hdeErrors :: NonEmpty e

httpJson
  :: (..., FromJSON a)
  => ...
  -> m (Response (Either HttpDecodeError String) a)
httpJson = httpDecode $ first pure . Aeson.eitherDecode -- NICE

httpDecode
  :: ...
  => (ByteString -> Either (NonEmpty e) a) -- Hmm, NonEmpty here
  -> Request
  -> m (Response (Either (HttpDecodeError e) a)) -- to e here... OK
```

But, uh,

```hs
instance Exception e => Exception (HttpDecodeError e) where
```

Doesn't work with `e ~ String`.

I thought of,

```hs
instance Show e => Exception (HttpDecodeError e) where
```

But I'll get double-quoting when I `show @String`.

Am I missing some trick or going down a wrong path here?

Reply:
Oh whoa, it looks like there's an `Exception` instance for `String`. That's unexpected, but means I can make this work.